crates/validation/src/header_validator.rs

```diff
@@ -207,26 +227,92 @@ impl HeaderValidator {
-    // https://github.com/ethereum/consensus-specs/blob/v0.11.1/ssz/merkle-proofs.md#generalized-merkle-tree-index
-    let hr_index = header.number % EPOCH_SIZE;
-    (EPOCH_SIZE * 2 * 2) + (hr_index * 2)
+    /// Verify that the execution block header is included in Deneb/Electra beacon block
+    #[must_use]
+    fn verify_post_deneb_execution_block_proof(
+        &self,
+        execution_header_hash: B256,
+        execution_block_proof: &[B256],
+        block_body_root: B256,
+    ) -> bool {
+        // BeaconBlock level:
+        // - 8 as there are 5 fields
+        // - 4 as index (pos) of field is 4
+        // let gen_index_top_level = (1 * 8 + 4) = 12
+        // BeaconBlockBody level:
+        // - 16 as there are 12 fields
+        // - 9 as index (pos) of field is 9
+        // let gen_index_mid_level = (gen_index_top_level * 16 + 9) = 201
+        // ExecutionPayload level:
+        // - 32 as there are 17 fields
+        // - 12 as pos of field is 12
+        // let gen_index = (gen_index_mid_level * 32 + 12) = 6444
+        let gen_index = 6444;
+
+        verify_merkle_proof(
+            execution_header_hash,
+            execution_block_proof,
+            execution_block_proof.len(),
+            gen_index,
+            block_body_root,
+        )
+    }
+
+    /// Verifies that pre-Capella Beacon Block root is included in the `HistoricalRoots`.
+    #[must_use]
+    fn verify_historical_roots_beacon_block_proof(
+        &self,
+        slot: u64,
+        beacon_block_root: B256,
+        beacon_block_proof: &BeaconBlockProofHistoricalRoots,
+    ) -> bool {
+        let block_root_index = slot % EPOCH_SIZE;
+        let gen_index = 2 * EPOCH_SIZE + block_root_index;
+        let historical_root_index = slot / EPOCH_SIZE;
+        let historical_root =
+            self.historical_roots_acc.historical_roots[historical_root_index as usize];
+
+        verify_merkle_proof(
+            beacon_block_root,
+            beacon_block_proof,
+            14,
+            gen_index as usize,
+            historical_root,
+        )
+    }
+
+    /// Verifies that post-Capella Beacon Block root is included in the [HistoricalSummaries].
+    #[must_use]
+    fn verify_historical_summaries_beacon_block_proof(
+        &self,
+        slot: u64,
+        beacon_block_root: B256,
+        beacon_block_proof: &BeaconBlockProofHistoricalSummaries,
+        historical_summaries: &HistoricalSummaries,
+    ) -> bool {
+        let block_root_index = slot % EPOCH_SIZE;
+        let gen_index = EPOCH_SIZE + block_root_index;
+        let historical_summary_index = (slot - CAPELLA_FORK_EPOCH * SLOTS_PER_EPOCH) / EPOCH_SIZE;
+        let historical_summary =
+            historical_summaries[historical_summary_index as usize].block_summary_root;
+
+        verify_merkle_proof(
+            beacon_block_root,
+            beacon_block_proof,
+            13,
+            gen_index as usize,
+            historical_summary,
```

Comment:
```suggestion
        let block_summary_root =
            historical_summaries[historical_summary_index as usize].block_summary_root;

        verify_merkle_proof(
            beacon_block_root,
            beacon_block_proof,
            13,
            gen_index as usize,
            block_summary_root,
```

maybe we should just call that variable `block_summary_root` as that is what it is

Reply:
I will do this:

```rust
let historical_summary = &historical_summaries[historical_summary_index as usize];
verify_merkle_proof(
    beacon_block_root,
    beacon_block_proof,
    13,
    gen_index as usize,
    historical_summary.block_summary_root,
)
```

Both name is correct and I think it's clearer (to me, block_summary_root is a bit vague).